src/views/tripPlannerForm.jsx

```diff
@@ -0,0 +1,305 @@
+      let contactList=[]
+      for(let i in contacts){
+        let contactData = getObject({objectList:this.props.user.contacts, key:"id", keyValue:contacts[i]})
+        if(contactData) contactList.push(contactData);
+      }
+      this.setState({contacts:contactList})
+    }
+
+    this.setState({
+      showMap:(this.state.points.length>0||this.state.paths.length>0),
+    })
+
+    this.fetchInReachData()
+  }
+
+  componentDidUpdate(prevProps){
+    //Update contact details after editing
+    if(prevProps.user.contacts!==this.props.user.contacts){
+      let updatedList=[]
+      for(let i in this.state.contacts){
+        let contact = getObject({objectList:this.props.user.contacts, key:"id", keyValue:this.state.contacts[i].id})
+        if(contact) updatedList.push(contact)
+      }
+      this.setState({contacts:updatedList})
+    }
+  }
+
+  fetchInReachData(){
+    if(this.props.user.profile && this.props.user.profile.mapshare_ID){
+      getKMLData({
+        mapshareID:this.props.user.profile.mapshare_ID, 
+        startDate:this.state.departTime, 
+        endDate:this.state.returnTime, 
+        onSuccess:(json)=>(this.setState({inReachData:parseInReachData(json)})), 
+        onFailure:()=>(console.log("No inreach data loaded"))
+      })
+    }
+  }
+
+  handleChange(event){
+    this.setState({[event.target.name]:event.target.value})
+  }
+
+  selectEmergencyContact(){
+    let title = <div>Add Emergency Contact</div>
+    let body = 
+    <div>
+      {this.props.user.contacts.length>0 ?
+        <ContactList 
+          contacts={this.props.user.contacts} 
+          app={this.props.app} 
+          actions={[
+            {label:"Use contact", action:(contact)=>{this.addEmergencyContact(contact); this.props.app.hideModal()}},
+          ]}
+        />
+        :
+        <p>You have no emergency contacts saved to your profile.</p>
+      }
+      <button 
+        className="btn text-left text-light btn-info btn-block my-2" 
+        onClick={()=>{
+          this.props.app.setModal(
+            <EditContact 
+              user={this.props.user} 
+              app={this.props.app}
+              onSuccess={(contact)=>{this.addEmergencyContact(contact)}}
+            />
+          )
+        }}
+        >+ New Contact
+      </button>
+    </div>
+    
+    this.props.app.setModal(<StandardModal title={title} body={body} hideModal={this.props.app.hideModal}/>)
+  }
+
+  addEmergencyContact(contact, mode="add"){
+    //Add or remove a contact from the current state
+    console.log("Adding contact", contact)
+    let existingContacts = this.state.contacts
+    let updatedContacts = []
+    for(let i in existingContacts){
+      if(contact===existingContacts[i]){
+        if(mode==="add"){
+          return
+        }else{
+          continue
+        }
+      }  
+      updatedContacts.push(existingContacts[i])
+    }
+    if(mode==="add"){updatedContacts.push(contact)}
+    this.setState({contacts:updatedContacts})
+  }
```

Comment:
Lets walk through this and see if we can make this a bit clearer. What does this do? It adds a contact to your existing contacts. 

Why is it taking an argument called `mode`? The function is called `addEmergencyContact`, why why would it sometimes take an arg called `remove`? 

I just want to make sure you understand the different between `for...of` loops for iterating over arrays, and `for...in` loops for iterating over properties in an object. I see lots of `for...in` which I think could just be `for...of`. 

I separated this into 2 functions so it is explicitly clear what each one does. I used some array methods to make things a bit more succinct. Check this out and let me know if you have any questions about this code! 

```js
  // Given a contact, add it to the list of contacts
  addEmergencyContact(contact){
    // Im assuming this is an array of contact objects...desctructing to deep copy, and not mutate state. 
    const existingContacts = [...this.state.contacts]

    // Check to see if this contact is already in the list
    // I dont know the structure of a contact, so I'm just going to assume it has an `id` property
    const isContactAlreadyAdded = existingContacts.find(c => c.id === contact.id)

    // If the contact is not already in the list, add it
    if (!isContactAlreadyAdded) {
      this.setState({contacts: [...existingContacts, contact]})
    }
  }

  deleteEmergencyContact(contact){
    // If the contact is in the list, remove it
    this.setState({contacts: this.state.contacts.filter(c => c.id !== contact.id)})
  }
```

Reply:
You're right, this function is a gongshow. Find, filter and map are going to be game changers, this is embarrassing I didn't even know about them.

Reply:
@mccutch There is A LOT you could learn about from languages to libraries to frameworks. Keeping up to date on the newest (circa 2017) parts of Javascript and React, are probably the most important things to keep up to date on. These are specifically called `Array Methods` and learning what all them do and how to use is properly the highest effort:reward thing you could learn.